srv/salt/_modules/osd.py

```diff
@@ -1455,22 +1472,44 @@ def remove(self):
 
         if self.force:
             log.warning("Forcing OSD removal")
-        else:
-            self.empty()
 
-        # Terminate
-        self.terminate()
+            # Terminate
+            self.terminate()
 
-        # Unmount filesystems
-        result = self.unmount()
-        if result:
-            return result
+            # Best effort depending on the reason for the forced removal
+            self.mark_destroyed()
+            update_destroyed(self._osd_disk(), self.osd_id)
+        else:
+            for func in [self.empty, self.terminate]:
+                msg = func()
+                if msg:
+                    log.error(msg)
+                    return msg
+
+            # Inform Ceph
+            #
+            # Consider this a hard requirement for graceful removal. If the
+            # OSD cannot be marked and recorded, stop the process.
+            if self.mark_destroyed():
+                msg = update_destroyed(self._osd_disk(), self.osd_id)
+                if msg:
+                    log.error(msg)
+                    return msg
+                log.debug("OSD {} marked and recorded".format(self.osd_id))
+            else:
+                msg = "Failed to mark OSD {} as destroyed".format(self.osd_id)
+                log.error(msg)
+                return msg
 
-        # Wipe partitions
-        self.wipe()
+        for func in [self.unmount, self.wipe, self.destroy]:
+            msg = func()
+            if msg:
+                log.error(msg)
+                return msg
+
+        # Remove grain
```

Comment:
Lately we ran into a situation where the remove.osd process supposedly was successful but actually did not run through completely which resulted in stale grains. We should somehow make people more aware that this command needs to be executed until it succeeds or bad things will happen :/